RFC: online raft consistency checker [ci skip]

vivekmenezes opened the pull request:
<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/4317)

<!-- Reviewable:end -->

vivekmenezes committed:
RFC: online raft consistency checker

tschottdorf commented on docs/RFCS/raft_consistency_checker.md:
```diff
@@ -0,0 +1,86 @@
+- Feature Name: Raft consistency checker
+- Status: draft
+- Start Date: 2016-02-10
+- Authors: Ben Darnell, Bram Grunier, Vivek Menezes
```
s/Grunier/Gruneir/

BramGruneir replied:
:+1:

tschottdorf commented on docs/RFCS/raft_consistency_checker.md:
```diff
@@ -0,0 +1,86 @@
+
+# Summary
+
+An online system that checks that periodically checks that range replicas
```
s/that perio/perio/

petermattis replied:
Too much checking going on: `s/that checks that/that/g`

tschottdorf commented on docs/RFCS/raft_consistency_checker.md:
```diff
@@ -0,0 +1,86 @@
+
+An online system that checks that periodically checks that range replicas
+are consistent; a snapshot for a range at a specific point in the raft log
+ has the same sha on all replicas.
```
nit: `s/^\s+//`, `s/sha/SHA/`

bdarnell replied:
Or s/sha/hash/gi; the specific hash used is an implementation detail (although it should be something from the SHA family; either sha512 from the go standard library or SHA3 from `x/crypto`)

tschottdorf commented on docs/RFCS/raft_consistency_checker.md:
```diff
@@ -0,0 +1,86 @@
+1. CHECK_CONSISTENCY_BEGIN (BEGIN): A replica on receiving this command
+immediately acknowledges it has executed it, and asynchronously computes the sha
+for the current snapshot. It continues executing commands while it computes the
+sha on the side. Once the sha is computed it compares the BEGIN command id along
```
just to be sure: command IDs don't exist any more. You could just use a nonce.

BramGruneir replied:
There's a bit missing here about creating a uuid for the request.  That's the payload that BEGIN sends.  Command ID should be replaced by that uuid.
We want to make sure the id is unique since any collision could result in a failure of consistency check.

bdarnell replied:
In addition to the UUID, both checksum requests should include a version number, and replicas ignore checksum requests with a version they don't understand.

tschottdorf commented on docs/RFCS/raft_consistency_checker.md:
```diff
@@ -0,0 +1,86 @@
+command snapshot on the leader. On receiving this command the replica saves the
+command id and sha supplied, and compares the sha against the one it eventually
+computes. Under the rare circumstances that it sees a difference, it will alert
+and euthanize the node.
```
sounds dangerous. If the leader is corrupted, every other node will kill itself? There needs to be some kind of quorum here.

tschottdorf replied:
I'm thinking

```go
// Pseudocode
func (r *Replica) AdminCheckConsistency() {
   result := r.proposeRaft(CheckConsistencyRequest{})
   shas := []string{result.SHA}
   // Get computed value from each Replica
   for _, replica := range replicas {
      // GetShaRequest returns the previously computed SHA (cached via the Raft execution timestamp).
      shas = append(shas, sendRPC(replica, GetShaRequest{Timestamp: result.Timestamp})).SHA)
   }
   if majoritySHA := computeMajoritySHA(shas); majoritySHA == "" {
     // We're in trouble; there's no majority of SHAs. This is a good reason to color the cluster red - no automatic solution possible.
   }
   for i, sha := range shas {
      if sha != majoritySHA {
         log.Warningf("replica %s is likely corrupt, suggesting orderly destruction", replicas[i]")
         sendRPC(replicas[i], ReplicaCorruptionRequest{})
      }
   }
}
```

BramGruneir replied:
This part was left a little vague.  At this point, auto-recovery probably needs more investigation and just looking for a majority might not be the right answer.

bdarnell replied:
If this kind of corruption occurs, it's impossible to tell which nodes (if any) are correct. It would be better to kill all the replicas than let the others limp along. The main goal of this checker is to raise alarms when it happens so we can start digging into logs and figure out what went wrong; we don't want recovery from this scenario to be too automatic. 

If we had some sort of quarantine mechanism, we could use that instead of killing the nodes, but `replicaCorruptionError` is not fully implemented yet.

tschottdorf commented on docs/RFCS/raft_consistency_checker.md:
```diff
@@ -0,0 +1,86 @@
+
+The above is implemented by introducing two raft commands:
+
+1. CHECK_CONSISTENCY_BEGIN (BEGIN): A replica on receiving this command
```
what are these commands? Can I picture them similar to other write commands offered through `(*Replica)` (like `Put`) or do they live somewhere else?

BramGruneir replied:
Yes, perhaps we should call them methods instead to match the proto.  Might be a good idea to do a pass through the code to ensure consistency of naming.

So two new methods.

bdarnell replied:
They're commands like `Put`. We generally use "command" instead of "method" for these (they're implemented by methods on `Replica`, but at all other layers of the stack they're not methods any more).

bdarnell replied:
These commands should be named something like `ComputeChecksum` and `VerifyChecksum`.

tschottdorf commented on docs/RFCS/raft_consistency_checker.md:
```diff
@@ -0,0 +1,86 @@
+computes. Under the rare circumstances that it sees a difference, it will alert
+and euthanize the node.
+
+Each range replica has the ability to keep track of multiple sha computations
```
if you key them by timestamp, it seems enough to hold on to the last one and error on mismatch (and the data is so small that actively cleaning it up may not make sense, especially if we're regularly checking). Or is there a use case I'm missing?

BramGruneir replied:
The worry is all of the scenarios list at the end.  But the cleanup is very minor.  But you're right, as soon as we have a positive match, all previous ones can be safely deleted.
What's not mentioned, but should be added is the frequency of this check.  Might be one a day or week in a production system, but it should be run multiple times in our acceptance tests.

tschottdorf commented on docs/RFCS/raft_consistency_checker.md:
```diff
@@ -0,0 +1,86 @@
+Noteworthy scenarios:
+
+1. The leader dies between BEGIN and CHECKSUM: The replicas, including the new
+leader, will continue computing the sha. The new leader will send a CHECKSUM
```
why not just let things be until the next queue run on the new leader begins with BEGIN again? That seems simpler and about as effective as introducing state and hence complexity.

BramGruneir replied:
Only reason would be the possible week long wait time between cycles.

vivekmenezes commented:
Review status: 0 of 1 files reviewed at latest revision, 8 unresolved discussions.

---

*[docs/RFCS/raft_consistency_checker.md, line 4 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KADKd95tMG41R1atWe7-r1-4:-KAFsop06TJ67gNeQGnc:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/969873aaa53bfe040cb875a6f22ce17bb450d6d1/docs/RFCS/raft_consistency_checker.md#L4)):*
done

---

*[docs/RFCS/raft_consistency_checker.md, line 10 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KADKd95tMG41R1atWe7-r1-10:-KAFspz5Y6SwZA-U2OXS:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/969873aaa53bfe040cb875a6f22ce17bb450d6d1/docs/RFCS/raft_consistency_checker.md#L10)):*
done

---

*[docs/RFCS/raft_consistency_checker.md, line 12 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KADKd95tMG41R1atWe7-r1-12:-KAFstSMwYw648rFpmbg:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/969873aaa53bfe040cb875a6f22ce17bb450d6d1/docs/RFCS/raft_consistency_checker.md#L12)):*
done

---

*[docs/RFCS/raft_consistency_checker.md, line 31 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KADKd95tMG41R1atWe7-r1-31:-KAFz4wOY5XT8hLhUk_9:46914075) ([raw file](https://github.com/cockroachdb/cockroach/blob/969873aaa53bfe040cb875a6f22ce17bb450d6d1/docs/RFCS/raft_consistency_checker.md#L31)):*
These commands will be new commands in roachpb.RequestUnion. Is that accurate?

---

*[docs/RFCS/raft_consistency_checker.md, line 34 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KADKd95tMG41R1atWe7-r1-34:-KAFwbviTxWzwbU1meN7:1744642273) ([raw file](https://github.com/cockroachdb/cockroach/blob/969873aaa53bfe040cb875a6f22ce17bb450d6d1/docs/RFCS/raft_consistency_checker.md#L34)):*
So Raft has no sequence numbering that I could use? I do see a Timestamp in the header.

---

*[docs/RFCS/raft_consistency_checker.md, line 42 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KADKd95tMG41R1atWe7-r1-42:-KAFxFyK44koctjQ7dBa:1290563680) ([raw file](https://github.com/cockroachdb/cockroach/blob/969873aaa53bfe040cb875a6f22ce17bb450d6d1/docs/RFCS/raft_consistency_checker.md#L42)):*
Thanks for the suggestion. I'm updating the design to reflect this.

---

*[docs/RFCS/raft_consistency_checker.md, line 44 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KADKd95tMG41R1atWe7-r1-44:-KAFvtnX87JNmXn99NgA:1535319932) ([raw file](https://github.com/cockroachdb/cockroach/blob/969873aaa53bfe040cb875a6f22ce17bb450d6d1/docs/RFCS/raft_consistency_checker.md#L44)):*
i think the concern here is if a leader reelection happens and a new leader sends another BEGIN. But you're right if we key by timestamp/sequence number it seems okay to always keep the latest one.

---

*[docs/RFCS/raft_consistency_checker.md, line 56 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KADKd95tMG41R1atWe7-r1-56:-KAFx3TRhRCHvyMqMxeN:-440672290) ([raw file](https://github.com/cockroachdb/cockroach/blob/969873aaa53bfe040cb875a6f22ce17bb450d6d1/docs/RFCS/raft_consistency_checker.md#L56)):*
The original thought is to run this once a day. Yeah, we probably don't want to do this.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Not yet updated the PR to reflect the latest changes. Hold on.

---

Review status: 0 of 1 files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-:-KAG9KkxApZSHfBpfT_i:-1073460063)*
<!-- Sent from Reviewable.io -->

petermattis commented on docs/RFCS/raft_consistency_checker.md:
```diff
@@ -0,0 +1,86 @@
+1. All replicas agree on the snapshot at which a sha is to be computed.
+2. A sha is computed on all replicas.
+3. The leader shares its sha with all replicas; a replica computing a
+different sha from the one supplied by the leader alerts and eutanizes
```
`s/eutanizes/euthanizes/g`

petermattis commented on docs/RFCS/raft_consistency_checker.md:
```diff
@@ -0,0 +1,86 @@
+
+# Alternatives
+
+A consistency checker that runs offline, or only in tests.
```
It might be useful to be able to force running of the consistency checker for a particular range or set of ranges. For example, in testing scenarios the background scan might not happen sufficiently frequently. You probably want to run a bunch of load against a system with nodes coming up and down, then be able to do something like `./cockroach check [<start-key>[,<end-key>]]` and have the consistency checks run on any ranges that intersect.

BramGruneir replied:
This is a great idea.

petermattis commented on docs/RFCS/raft_consistency_checker.md:
```diff
@@ -0,0 +1,86 @@
+
+# Drawbacks
+
+None.
```
There is a potential impact on performance. That's not to say we shouldn't do this, but that there is a drawback.

bdarnell replied:
Another potential drawback: a bug in the consistency checker could cause nodes to die when there is no actual consistency problem.

bdarnell commented on docs/RFCS/raft_consistency_checker.md:
```diff
@@ -0,0 +1,86 @@
+The above is implemented by introducing two raft commands:
+
+1. CHECK_CONSISTENCY_BEGIN (BEGIN): A replica on receiving this command
+immediately acknowledges it has executed it, and asynchronously computes the sha
```
The computation needs to be specified in more detail. I think it needs to cover everything returned by `replicaDataIterator` with two exceptions: the raft `HardState` and any log entries with index greater than `appliedIndex` (at the time we took the snapshot).

bdarnell commented on docs/RFCS/raft_consistency_checker.md:
```diff
@@ -0,0 +1,86 @@
+Nothing.
+7. A new replica is added between BEGIN and CHECKSUM: The new replica receives
+the CHECKSUM and ignores it.
+8. A replica is removed between BEGIN and CHECK: Nothing.
```
What about splits and merges?

tschottdorf commented:
Reviewed 1 of 1 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[docs/RFCS/raft_consistency_checker.md, line 42 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KADKd95tMG41R1atWe7-r1-42:-KAGoj5hNpTV_nh47d4k:1615088222) ([raw file](https://github.com/cockroachdb/cockroach/blob/969873aaa53bfe040cb875a6f22ce17bb450d6d1/docs/RFCS/raft_consistency_checker.md#L42)):*
I think that the RFC should be limited in scope to report anomalies only. We could have an AdminCommand that sends out the ComputeConsistency and GetConsistency, returning to the caller of AdminCommand the nodes' SHAs. Later a queue can call that and do cluster maintenance with the data. That keeps the scope here to what we actually need - something to call from an acceptance test that tells us whether all replicas have the same state.

---

*[docs/RFCS/raft_consistency_checker.md, line 44 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KADKd95tMG41R1atWe7-r1-44:-KAGpLWY35EIVKBwyoLO:1396932918) ([raw file](https://github.com/cockroachdb/cockroach/blob/969873aaa53bfe040cb875a6f22ce17bb450d6d1/docs/RFCS/raft_consistency_checker.md#L44)):*
Just using a UUID is fine (it's extremely unlikely but possible that two consecutive leaders chose the same timestamp) and keeping the last result (with the UUID). If a node didn't see the requested nonce last, it'll just return an error.
As written above, it seems to me that the second command shouldn't include verification, but just request the result so that the caller can act on the aggregate output.

---

*[docs/RFCS/raft_consistency_checker.md, line 56 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KADKd95tMG41R1atWe7-r1-56:-KAGpxGeC4J2KVSnJMGv:-977904445) ([raw file](https://github.com/cockroachdb/cockroach/blob/969873aaa53bfe040cb875a6f22ce17bb450d6d1/docs/RFCS/raft_consistency_checker.md#L56)):*
With the focus on an `AdminCollectChecksums` command (or whatever the name) as suggested above, nothing interesting happens in any of the noteworthy scenarios.

---

*[docs/RFCS/raft_consistency_checker.md, line 74 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KADKd95tMG41R1atWe7-r1-74:-KAGq8yFP2iat06cZbtF:-1930048689) ([raw file](https://github.com/cockroachdb/cockroach/blob/969873aaa53bfe040cb875a6f22ce17bb450d6d1/docs/RFCS/raft_consistency_checker.md#L74)):*
Splits and merges should clear out the saved state (so that compute/get-sha can only be carried out successfully as long as the range hasn't changed its boundary).

---

*[docs/RFCS/raft_consistency_checker.md, line 78 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KADKd95tMG41R1atWe7-r1-78:-KAGqiuLgV-DzVgLxSfl:-717120986) ([raw file](https://github.com/cockroachdb/cockroach/blob/969873aaa53bfe040cb875a6f22ce17bb450d6d1/docs/RFCS/raft_consistency_checker.md#L78)):*
The Raft command can execute on a SNAPSHOT and doesn't actually block processing. Looks like this should be doable in a way  that doesn't incur a terrible perf hit, but I'm in favor of working up something that the acceptance tests can use for now.

---

*[docs/RFCS/raft_consistency_checker.md, line 82 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KADKd95tMG41R1atWe7-r1-82:-KAGr1hD23fu1zcZ2xYP:-1524852769) ([raw file](https://github.com/cockroachdb/cockroach/blob/969873aaa53bfe040cb875a6f22ce17bb450d6d1/docs/RFCS/raft_consistency_checker.md#L82)):*
the admin command I suggested is a good way to do this for the acceptance tests. And I like Pete's idea of shelling out most of the actual hash computation so that a cli command can reuse it against keyranges on a per-store basis.

---

*[docs/RFCS/raft_consistency_checker.md, line 39 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KAGs0Go47nbxvBoY7js:-KAGs0Go47nbxvBoY7jt:284598802) ([raw file](https://github.com/cockroachdb/cockroach/blob/07ff8b5ecbb2933d446f423134a0f30e4bab5335/docs/RFCS/raft_consistency_checker.md#L39)):*
The scanner could be a future development. It's not immediately necessary since it's better for the acceptance tests to have direct control.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 0 of 1 files reviewed at latest revision, 13 unresolved discussions.

---

*[docs/RFCS/raft_consistency_checker.md, line 10 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KADKd95tMG41R1atWe7-r1-10:-KAGr_VGbG1RCYj_90UL:-467915377) ([raw file](https://github.com/cockroachdb/cockroach/blob/969873aaa53bfe040cb875a6f22ce17bb450d6d1/docs/RFCS/raft_consistency_checker.md#L10)):*
fixed

---

*[docs/RFCS/raft_consistency_checker.md, line 12 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KADKd95tMG41R1atWe7-r1-12:-KAGrafRl5vXx0jPA-TX:1819922056) ([raw file](https://github.com/cockroachdb/cockroach/blob/969873aaa53bfe040cb875a6f22ce17bb450d6d1/docs/RFCS/raft_consistency_checker.md#L12)):*
I've left it as SHA

---

*[docs/RFCS/raft_consistency_checker.md, line 26 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KADKd95tMG41R1atWe7-r1-26:-KAGreSKr6rBlpTef4AS:-467915377) ([raw file](https://github.com/cockroachdb/cockroach/blob/969873aaa53bfe040cb875a6f22ce17bb450d6d1/docs/RFCS/raft_consistency_checker.md#L26)):*
fixed

---

*[docs/RFCS/raft_consistency_checker.md, line 31 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KADKd95tMG41R1atWe7-r1-31:-KAGzYWY-ina5917M-Lm:676635175) ([raw file](https://github.com/cockroachdb/cockroach/blob/969873aaa53bfe040cb875a6f22ce17bb450d6d1/docs/RFCS/raft_consistency_checker.md#L31)):*
Done

---

*[docs/RFCS/raft_consistency_checker.md, line 32 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KADKd95tMG41R1atWe7-r1-32:-KAIE30efQFl_1SeDY-b:-1205589748) ([raw file](https://github.com/cockroachdb/cockroach/blob/969873aaa53bfe040cb875a6f22ce17bb450d6d1/docs/RFCS/raft_consistency_checker.md#L32)):*
Added this comment to the doc.

---

*[docs/RFCS/raft_consistency_checker.md, line 34 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KADKd95tMG41R1atWe7-r1-34:-KAIELLWRI5VgokD3llP:-1281442526) ([raw file](https://github.com/cockroachdb/cockroach/blob/969873aaa53bfe040cb875a6f22ce17bb450d6d1/docs/RFCS/raft_consistency_checker.md#L34)):*
Do they really need a version number. I thought commands are ordered in raft and a VerifyChecksum command can be paired with the latest ComputeChecksum command.

---

*[docs/RFCS/raft_consistency_checker.md, line 42 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KADKd95tMG41R1atWe7-r1-42:-KAIEojVOdSzoRxLVqyL:-1259200153) ([raw file](https://github.com/cockroachdb/cockroach/blob/969873aaa53bfe040cb875a6f22ce17bb450d6d1/docs/RFCS/raft_consistency_checker.md#L42)):*
The RFC is limited in scope to only check for anomalies in consistency and report them. Tests can be configured to crash when they see anomalies. I'm going to add an admin command to compute consistency checksums on the data over ranges.

---

*[docs/RFCS/raft_consistency_checker.md, line 56 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KADKd95tMG41R1atWe7-r1-56:-KAIKWhp2matERjf4bdV:-948854872) ([raw file](https://github.com/cockroachdb/cockroach/blob/969873aaa53bfe040cb875a6f22ce17bb450d6d1/docs/RFCS/raft_consistency_checker.md#L56)):*
I've kept the periodic check along with a check that can be run in an acceptance test.

---

*[docs/RFCS/raft_consistency_checker.md, line 74 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KADKd95tMG41R1atWe7-r1-74:-KAIKg3efjJ1OcOj8C2Q:-2033898451) ([raw file](https://github.com/cockroachdb/cockroach/blob/969873aaa53bfe040cb875a6f22ce17bb450d6d1/docs/RFCS/raft_consistency_checker.md#L74)):*
added this to the doc.

---

*[docs/RFCS/raft_consistency_checker.md, line 78 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KADKd95tMG41R1atWe7-r1-78:-KAGs89P1j3ZZU48L7EL:-472701157) ([raw file](https://github.com/cockroachdb/cockroach/blob/969873aaa53bfe040cb875a6f22ce17bb450d6d1/docs/RFCS/raft_consistency_checker.md#L78)):*
added

---

*[docs/RFCS/raft_consistency_checker.md, line 82 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KADKd95tMG41R1atWe7-r1-82:-KAIKkW_uI2cxb7iNZSs:-1866581810) ([raw file](https://github.com/cockroachdb/cockroach/blob/969873aaa53bfe040cb875a6f22ce17bb450d6d1/docs/RFCS/raft_consistency_checker.md#L82)):*
Thanks for the suggestions folks. I've added these to the doc.

---

*[docs/RFCS/raft_consistency_checker.md, line 39 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KAGs0Go47nbxvBoY7js:-KAIFxk6xXeJETNO-p19:349480573) ([raw file](https://github.com/cockroachdb/cockroach/blob/07ff8b5ecbb2933d446f423134a0f30e4bab5335/docs/RFCS/raft_consistency_checker.md#L39)):*
I'd like to run this in every cluster and report consistency issues.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Ready for another review. Thanks!

---

Review status: 0 of 1 files reviewed at latest revision, 13 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-:-KAIKsuCSb4jXugTnZVj:1862545688)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/raft_consistency_checker.md, line 34 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KADKd95tMG41R1atWe7-r1-34:-KAJ7ncgUtX1zFDpyM-y:478846923) ([raw file](https://github.com/cockroachdb/cockroach/blob/969873aaa53bfe040cb875a6f22ce17bb450d6d1/docs/RFCS/raft_consistency_checker.md#L34)):*
I think @bdarnell is worrying about clusters in which various cockroach versions are running concurrently.

---

*[docs/RFCS/raft_consistency_checker.md, line 42 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KADKd95tMG41R1atWe7-r1-42:-KAJ4PxUeZhmyl92O7H6:1911344529) ([raw file](https://github.com/cockroachdb/cockroach/blob/969873aaa53bfe040cb875a6f22ce17bb450d6d1/docs/RFCS/raft_consistency_checker.md#L42)):*
I'm suggesting that the better interface is one in which the second command isn't `VerifyChecksum`, but `CollectChecksums` because that is more versatile than the design presented here (where each replica gets handed a checksum and then, in isolation, needs to figure out what its course of action is).

For example, in an acceptance test, I want to call something like
```go
// runs VerifyChecksum through Raft and then collects all SHAs via CollectChecksums.
shas := AdminComputeSHAs(roachpb.KeyMin, roachpb.KeyMax)
// Check if all SHAs are the same, otherwise die.
```

and a scanner queue would similarly use `shas` to warn/take more intelligent action later. That seems like a good building block for any type of functionality we could possibly want.

---

*[docs/RFCS/raft_consistency_checker.md, line 88 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KAJ6ZR1i9z9YXeOvdQu:-KAJ6ZR1i9z9YXeOvdQv:1611366562) ([raw file](https://github.com/cockroachdb/cockroach/blob/6826e0462071713d8ec4c56580065dee11687f2f/docs/RFCS/raft_consistency_checker.md#L88)):*
I don't understand the first part and the language sounds optional.

Suggestion: On a range split or merge, all participating Replicas clear out any computed SHAs (or SHAs which are still being computed).

Thinking about this more though, I think with a UUID there's nothing that could go wrong. The only enemy here is routing the second RPC to the wrong range (due to split/merge), but then that range wouldn't have a matching UUID. Still good to avoid, I suppose.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/raft_consistency_checker.md, line 34 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KADKd95tMG41R1atWe7-r1-34:-KALE4cEOBo6oTMdPJ2T:-231745067) ([raw file](https://github.com/cockroachdb/cockroach/blob/969873aaa53bfe040cb875a6f22ce17bb450d6d1/docs/RFCS/raft_consistency_checker.md#L34)):*
Good point. I've added a UUID and a version.

---

*[docs/RFCS/raft_consistency_checker.md, line 42 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KADKd95tMG41R1atWe7-r1-42:-KALJWnhHoj4JLPcZuvY:-499145013) ([raw file](https://github.com/cockroachdb/cockroach/blob/969873aaa53bfe040cb875a6f22ce17bb450d6d1/docs/RFCS/raft_consistency_checker.md#L42)):*

I've replaced VerifyChecksum with CollectChecksum. The admin command will initially be just a check to verify that SHAs on all replicas for local leader range replicas are consistent. The command can be applied on every node in the cluster in the acceptance tests. 

---

*[docs/RFCS/raft_consistency_checker.md, line 44 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KADKd95tMG41R1atWe7-r1-44:-KALmmtsVVkHK-Y441yk:677042770) ([raw file](https://github.com/cockroachdb/cockroach/blob/969873aaa53bfe040cb875a6f22ce17bb450d6d1/docs/RFCS/raft_consistency_checker.md#L44)):*
SGTM

---

*[docs/RFCS/raft_consistency_checker.md, line 88 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KAJ6ZR1i9z9YXeOvdQu:-KALsMSC-Mm4IKCIXlD_:1922924238) ([raw file](https://github.com/cockroachdb/cockroach/blob/6826e0462071713d8ec4c56580065dee11687f2f/docs/RFCS/raft_consistency_checker.md#L88)):*
On a range split/merge we invalidate the outstanding ComputeChecksum UUID on the non-leader replica and return a NOT_COMPUTED status in the CollectChecksum response. 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Do let me know if there is any more feedback on this RFC. Thanks

On Fri, Feb 12, 2016, 1:21 AM Tobias Schottdorf <notifications@github.com>
wrote:

> Reviewed 1 of 1 files at r4.
> Review status: all files reviewed at latest revision, 11 unresolved
> discussions, some commit checks failed.
> ------------------------------
>
> *docs/RFCS/raft_consistency_checker.md, line 34 [r1]
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KADKd95tMG41R1atWe7-r1-34:-KAJ7ncgUtX1zFDpyM-y:478846923>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/969873aaa53bfe040cb875a6f22ce17bb450d6d1/docs/RFCS/raft_consistency_checker.md#L34>):*
> I think @bdarnell <https://github.com/bdarnell> is worrying about
> clusters in which various cockroach versions are running concurrently.
> ------------------------------
>
> *docs/RFCS/raft_consistency_checker.md, line 42 [r1]
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KADKd95tMG41R1atWe7-r1-42:-KAJ4PxUeZhmyl92O7H6:1911344529>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/969873aaa53bfe040cb875a6f22ce17bb450d6d1/docs/RFCS/raft_consistency_checker.md#L42>):*
>
> I'm suggesting that the better interface is one in which the second
> command isn't VerifyChecksum, but CollectChecksums because that is more
> versatile than the design presented here (where each replica gets handed a
> checksum and then, in isolation, needs to figure out what its course of
> action is).
>
> For example, in an acceptance test, I want to call something like
>
> // runs VerifyChecksum through Raft and then collects all SHAs via CollectChecksums.shas := AdminComputeSHAs(roachpb.KeyMin, roachpb.KeyMax)// Check if all SHAs are the same, otherwise die.
>
> and a scanner queue would similarly use shas to warn/take more
> intelligent action later. That seems like a good building block for any
> type of functionality we could possibly want.
> ------------------------------
>
> *docs/RFCS/raft_consistency_checker.md, line 88 [r4]
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KAJ6ZR1i9z9YXeOvdQu:-KAJ6ZR1i9z9YXeOvdQv:1611366562>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/6826e0462071713d8ec4c56580065dee11687f2f/docs/RFCS/raft_consistency_checker.md#L88>):*
> I don't understand the first part and the language sounds optional.
>
> Suggestion: On a range split or merge, all participating Replicas clear
> out any computed SHAs (or SHAs which are still being computed).
>
> Thinking about this more though, I think with a UUID there's nothing that
> could go wrong. The only enemy here is routing the second RPC to the wrong
> range (due to split/merge), but then that range wouldn't have a matching
> UUID. Still good to avoid, I suppose.
> ------------------------------
>
> *Comments from the review on Reviewable.io
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/4317>*
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/4317#issuecomment-183197161>
> .
>

JackKrupansky commented:
I'm coming from a Cassandra background, so please discount anything that is based on design differences between Cassandra and CockroachDB, although they are both LSM and both dealing with a lot of the same distributed data issues. My notes...

1. +1 for a true "offline", last-resort checker. For testing it makes solid sense.

2. Best if the checker is as as separate from the normal processing as possible so that bugs in normal range processing code don't simultaneously infect the checker.

3. Be clear as to whether a typical installation is going to be expected to frequently/occasionally run the checker, like daily, weekly, or whatever. Or... is this really only intended for in-house/dev use, not normal production clusters out in the wild?

4. Can the check reasonably be run on an active cluster, while write transactions are occurring? I mean, won't the SHA of one replica frequently be out of sync with other replicas which simply haven't yet finished their replica commit of the transaction and will likely very shortly? Sure, the checker could skip or adjust for these in-progress ranges, but that would see to add a lot of complexity to the checker, which again should be as simple and rock-solid reliable (low-complexity) as possible.

5. Can the checker depend on some unspecified mechanism for freezing or snapshotting the cluster so that in-progress transactions don't cause temporary check failures?

6. Is the range replica SHA computed on the raw KV store keys and values? (see next question)

7. Does the range replica SHA include or exclude write intents for pending transactions - or even committed transactions which are cleaning up intents after the commit, or maybe even cases where the node went down after the commit but before intent cleanup completed and then came up (when a manual check might be a very advisable admin function)?

8. It has never been clear to me how replicas get synced after they have been offline while a range quorum commits for a transaction and then the replica comes back online. Cassandra has hinted-hand off. But even then, if a node is offline for too long, like an hour, the hints get dropped and then the operator must manually run a repair to force the node to stream updated data from one of the quorum nodes that had committed the data. What is the comparable process in CockroachDB, and how does the checker interact or contrast with that normal out-of-quorum update process.

Again, my apologies if any of the Stone Age Cassandra distributed database concepts that my experience is based on don't apply to the more modern CockroachDB technology.

bdarnell commented on docs/RFCS/raft_consistency_checker.md:
```diff
@@ -0,0 +1,114 @@
+of time to compute the SHA. If the replica has not computed the SHA it returns
+a NOT_COMPUTED status. In tests this command will block until the checksum is
+computed. If this command never arrives the computed SHA is deleted the next
+time a new ComputeChecksum is seen. 
```
If we have both a manual and automatic mode then I think we need to keep multiple checksums around. We don't want the background scan to disrupt a manually-initiated scan.

bdarnell commented on docs/RFCS/raft_consistency_checker.md:
```diff
@@ -0,0 +1,114 @@
+replica receives the CollectChecksum and returns NOT_COMPUTED.
+8. A replica is removed between ComputeChecksum and CollectChecksum: Nothing.
+9. A range splits or merges with another range after ComputeChecksum: In both
+scenarios the UUID will be invalidated and NOT_COMPUTED will be returned by
```
We don't necessarily need to invalidate the checksum on a split; it could be carried forward on the left-hand side to the new range. But this will leave gaps in the sequence; the right-hand side that occurs during a consistency check scan will not be checked unless we have special handling for this. If this is the daily background scan that's probably fine; it'll be picked up in the next cycle. But in a manual check, are these gaps going to be common enough that we need to go back and re-scan ranges that split as we check them? I'm not sure.

bdarnell commented:
LGTM

tschottdorf commented:
LGTM. Do you have a plan for addressing the individual replicas (not required for the RFC, just interested) for `CollectChecksum`?

---

Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 17 unresolved discussions.

---

*[docs/RFCS/raft_consistency_checker.md, line 28 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KARJV9PJYbxrbKpeHtI:-KARJV9PJYbxrbKpeHtJ:1278352943) ([raw file](https://github.com/cockroachdb/cockroach/blob/4188da8c718c43669e199e29ecc68e036966f11f/docs/RFCS/raft_consistency_checker.md#L28)):*
a SHA different from what? All replicas could return a different SHA. Probably want to say: reports if not all returned SHAs match.

---

*[docs/RFCS/raft_consistency_checker.md, line 37 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KARK5wK6tBTYWwvLL0V:-KARK5wK6tBTYWwvLL0W:1217314551) ([raw file](https://github.com/cockroachdb/cockroach/blob/4188da8c718c43669e199e29ecc68e036966f11f/docs/RFCS/raft_consistency_checker.md#L37)):*
s/The command.*computing the SHA/The command contains a UUID under which the result is to be made available for subsequent querying via CollectChecksum below. The version number is checked by the executing replica against its own version number, and an error returned on mismatch/

---

*[docs/RFCS/raft_consistency_checker.md, line 39 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KARK_sto29fuver4ntA:-KARK_suIUUTUsBokhIN:328103825) ([raw file](https://github.com/cockroachdb/cockroach/blob/4188da8c718c43669e199e29ecc68e036966f11f/docs/RFCS/raft_consistency_checker.md#L39)):*
s/a UUID/the UUID/

---

*[docs/RFCS/raft_consistency_checker.md, line 41 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KARKlmKRfAyJ74JwyLa:-KARKlmKRfAyJ74JwyLb:-1236994475) ([raw file](https://github.com/cockroachdb/cockroach/blob/4188da8c718c43669e199e29ecc68e036966f11f/docs/RFCS/raft_consistency_checker.md#L41)):*
I think for symmetry the leader should not "compute its own sha" but just send the same command it sends to other replicas to itself as well.

---

*[docs/RFCS/raft_consistency_checker.md, line 46 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KADKd95tMG41R1atWe7-r5-46:-KARLAeq5mtJxbSGU0_U:2129652735) ([raw file](https://github.com/cockroachdb/cockroach/blob/4188da8c718c43669e199e29ecc68e036966f11f/docs/RFCS/raft_consistency_checker.md#L46)):*
that's true. To avoid complexity from that, it should be easy enough to have a mutex-protected `map[UUID]result` and insertion code

```go
mu.Lock()
defer mu.Unlock()
if _, ok := m[uuid]; ok {
  return
}
m[uuid] = result
stopper.RunAsyncTask(func() {
  mu.Lock()
  defer mu.Unlock()
  select {
    case stopper.ShouldDrain():
    case <-time.After(5*time.Minute):
  }
  delete(m, uuid)
})
```

> In tests this command will block until the checksum is
computed

s/tests/manual invocations (e.g. tests or possibly a later `cli` command)/

---

*[docs/RFCS/raft_consistency_checker.md, line 66 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KAROACD1WiTh3Y4aQVC:-KAROACD1WiTh3Y4aQVD:-455514669) ([raw file](https://github.com/cockroachdb/cockroach/blob/4188da8c718c43669e199e29ecc68e036966f11f/docs/RFCS/raft_consistency_checker.md#L66)):*
s/KeyMax\]/KeyMax)/

---

*[docs/RFCS/raft_consistency_checker.md, line 89 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KADKd95tMG41R1atWe7-r5-89:-KAROouX1v6FgRtY_Af7:-1113480659) ([raw file](https://github.com/cockroachdb/cockroach/blob/4188da8c718c43669e199e29ecc68e036966f11f/docs/RFCS/raft_consistency_checker.md#L89)):*
Propagating to the left part on a Split makes sense. My caution here is mostly due to the fact that we currently don't have a good mechanism for addressing individual replicas. I'd wait with addressing the problems with gaps until they become relevant. It doesn't seem so bad to have a gap upon split: after all, the data in the gap was just copied from the left side after we started checking consistency, so we still see all of the data and the window in which corruption could have happened is minimal. Merges are worse in that respect, because data we would have wanted to scan effectively isn't queried. But that manifests as a nonexistent range on CollectChecksum, and any kind of cli tool would just query that missing range separately.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-:-KARQgyQMa0_Bm2IrjBn:530537667)*
<!-- Sent from Reviewable.io -->

BramGruneir commented on docs/RFCS/raft_consistency_checker.md:
```diff
@@ -0,0 +1,114 @@
+command will contain a UUID and a version number identifying the method
+used in computing the SHA. The SHA is reported back to the leader through the
+next command.
+2. CollectChecksum: The leader will supply a UUID for the ComputeChecksum
```
Shouldn't collect checksum include the checksum calculated by the leader?  So we can reply and possibly panic on a follower replica?

tschottdorf replied:
the leader is also a Replica, so it also collects its own.

BramGruneir replied:
Of course, but it might be nice to know we have a mismatch when any replica receives the collect checksum command and the checksums don't match.

tschottdorf replied:
but that doesn't seem useful because that only allows the replica to compare itself to the leader, which isn't more likely to be correct than itself. I think the idea is that the leader collects all the checksum and can than take an informed action.

BramGruneir replied:
I think there is value in a replica knowing that there is a conflict.  If you have an extremely slow store, say due to bad sectors, unless the leader keeps polling and there are no leadership changes, it might be missed entirely.  Unlikely, but possible.  It's a small amount of info to send as part of the command and in the least allows us to throw a warning that something is wrong.

BramGruneir commented:
LGTM

vivekmenezes commented:
Review status: 0 of 1 files reviewed at latest revision, 18 unresolved discussions.

---

*[docs/RFCS/raft_consistency_checker.md, line 28 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KARJV9PJYbxrbKpeHtI:-KAXeKDvjVFPVN0xzv8r:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/4188da8c718c43669e199e29ecc68e036966f11f/docs/RFCS/raft_consistency_checker.md#L28)):*
done

---

*[docs/RFCS/raft_consistency_checker.md, line 37 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KARK5wK6tBTYWwvLL0V:-KAXfk7xbtOjSSWPBdRY:676635175) ([raw file](https://github.com/cockroachdb/cockroach/blob/4188da8c718c43669e199e29ecc68e036966f11f/docs/RFCS/raft_consistency_checker.md#L37)):*
Done

---

*[docs/RFCS/raft_consistency_checker.md, line 39 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KADKd95tMG41R1atWe7-r5-39:-KAXj4HzycfX_o3ZNqOL:-1803417269) ([raw file](https://github.com/cockroachdb/cockroach/blob/4188da8c718c43669e199e29ecc68e036966f11f/docs/RFCS/raft_consistency_checker.md#L39)):*
We plan to report errors at the leader including NOT_COMPUTED status errors. I'm not adding in the leader SHA into the request for now.

---

*[docs/RFCS/raft_consistency_checker.md, line 39 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KARK_sto29fuver4ntA:-KAXjiRwrfsaJcAwqZ0M:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/4188da8c718c43669e199e29ecc68e036966f11f/docs/RFCS/raft_consistency_checker.md#L39)):*
done

---

*[docs/RFCS/raft_consistency_checker.md, line 41 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KARKlmKRfAyJ74JwyLa:-KAXjpOvaXOJMJttsyEM:676635175) ([raw file](https://github.com/cockroachdb/cockroach/blob/4188da8c718c43669e199e29ecc68e036966f11f/docs/RFCS/raft_consistency_checker.md#L41)):*
Done

---

*[docs/RFCS/raft_consistency_checker.md, line 46 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KADKd95tMG41R1atWe7-r5-46:-KAXnmijmz_0Jqzhpr1M:-2137399250) ([raw file](https://github.com/cockroachdb/cockroach/blob/4188da8c718c43669e199e29ecc68e036966f11f/docs/RFCS/raft_consistency_checker.md#L46)):*
Thanks

---

*[docs/RFCS/raft_consistency_checker.md, line 66 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KAROACD1WiTh3Y4aQVC:-KAXnxp6MEE_2Ubvhpm1:1914133186) ([raw file](https://github.com/cockroachdb/cockroach/blob/4188da8c718c43669e199e29ecc68e036966f11f/docs/RFCS/raft_consistency_checker.md#L66)):*
Hmmm yes. Very precise indeed. Thanks.

---

*[docs/RFCS/raft_consistency_checker.md, line 89 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317#-KADKd95tMG41R1atWe7-r5-89:-KAXqafIKtlsQ67TAGbb:-1624069688) ([raw file](https://github.com/cockroachdb/cockroach/blob/4188da8c718c43669e199e29ecc68e036966f11f/docs/RFCS/raft_consistency_checker.md#L89)):*
I'd like to discuss this offline because I'm not familiar with the specifics of range split and merge.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4317)*
<!-- Sent from Reviewable.io -->
